### ISSUE TITLE
CoreDNS customization options.

### ISSUE THREAD

**Issue by captncraig** [kontena/pharos-cluster#1082]
I have a use case where we want to have cluster dns available outside the cluster for our other dns infrastructure to delegate to. I wanted to run coredns:

- As a daemonset on the master nodes only.
- Using hostnetwork to bind to 53 on the masters.
- Using a custom configmap (custom domains, a few additional options and so forth)

That gave me a fixed location on the correct ports to make sure dns is always available anywhere. But I couldn't figure out how to customize the existing coredns deployment to act that way (although that would be nice). I settled on using the existing deployment for in-cluster dns, and my own daemonset for external use. I would love to combine the two if possible, but the dual setup seems to work ok.

Until today. I learned I need to remove the `loadbalance` from the in-cluster deployment's config. (Upstream is an AD dns server that returns records in order with the local subnet first. We need to preserve that order for... reasons.) So I need to figure out now how to customize that config map. I can update it manually to make it work, but that is not a durable solution going forward.

So... What are my options for customizing coredns? A few possible options I can think of:

- Add individual config options for each aspect of coredns that anyone asks to have customizable.
- Allow some option to bring a full custom configmap.
- Some option to disable the built-in coredns deployment entirely, and I just bring my own (possibly with a custom addon). 

The more I think about it, maybe the third option is the way to go for simplicity, but I am not sure quite how all of the existing moving pieces work together. It looks like maybe kubeadm is creating the base coredns deployment and pharos is customizing it a little bit? I'm not sure what it looks like to bring my own completely, especially with things like configuring the cluster to use it as the native dns resolver in the cluster.

I'm sorry to rant so much... any thoughts?

**Comment by captncraig** [comment]
Ok, I've experimented a bit and learned a lot. Here's a solution I've been able to make work:

1. Make my own coredns configmap and daemonset and get them working as I want.
2. If the pods have `k8s-app: kube-dns`, they will get added to the existing `kube-dns` service, and be used by the rest of the cluster.
3. I can then delete the kubeadm/pharos created coredns configmap and deployment, and everything seems to keep working.

I will probably make a local addon to do step 3 so it doesn't keep popping up every time we run pharos. If it makes sense to have a "don't deploy coredns" option in pharos, that might be nice, but I also don't want the kube-dns service to disappear, so maybe best not poke that particular hornet's nest. 

I would be ok closing this, or if you think any changes make sense, I can leave it open. Not urgent at this point in any case.

**Comment by captncraig** [comment]
I take it back. After deleting the coredns deployment, the next run of pharos fails with 

```
Patching coredns deployment with 2 replicas (max-surge 1, max-unavailable 1)...
    got error (K8s::Error::NotFound): PATCH /apis/extensions/v1beta1/namespaces/kube-system/deployments/coredns => HTTP 404 Not Found: deployments.extensions "coredns" not found
```

so I guess maybe I do need some level of support to do what I am trying to do.

**Comment by captncraig** [comment]
I attempted to set dns_replicas to `0` as a hack, but the config validator didn't like that either.

**Comment by captncraig** [comment]
Ok, I think I figured out a way to do this with a custom addon. [gist here](https://gist.github.com/captncraig/37cd46a8ebd1c586d601b706461f1b2c).

The idea is: 

- I can overwrite the `coredns` configmap with my own (except [this bug](https://github.com/kontena/k8s-client/pull/108) gave me a headache doing that with `apply_resources`.
- I can create my daemonset of coredns in addition to the standard deployment (which I scale down to 1 pod). All of the pods will serve in-cluster queries, but that seems ok. The deployment and the daemonset will share the patched configmap.
- Because the configmap is patched as an addon, I can be reasonably confident it will not be overwritten at some future time without giving the patch a chance to re-apply itself. 

I think I am happy with this solution in the end. And I learned some strategies for patching some of the "built-in" objects that I think I can use for some other things.
